pkg/operator/ceph/webhook-config.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2022 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package operator
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	api "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	v1 "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	admv1 "k8s.io/api/admissionregistration/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	issuerName        = "selfsigned-issuer"
+	certificateName   = "rook-admission-controller-cert"
+	webhookConfigName = "rook-ceph-webhook"
+)
+
+func createIssuer(ctx context.Context, clusterdContext *clusterd.Context) (*api.Issuer, error) {
+	logger.Infof("Fetching Issuer %s/%s.", namespace, issuerName)
+	issuer, err := clusterdContext.Admission.Issuers(namespace).Get(ctx, issuerName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Infof("Creating Issuer %s/%s.", namespace, issuerName)
+			issuer = &api.Issuer{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      issuerName,
+					Namespace: namespace,
+				},
+				Spec: api.IssuerSpec{
+					IssuerConfig: api.IssuerConfig{
+						SelfSigned: &api.SelfSignedIssuer{},
+					},
+				},
+			}
+			issuers, _ := clusterdContext.Admission.Issuers(namespace).Create(ctx, issuer, metav1.CreateOptions{})
+			sleepTime := 1
+			attempts := 10
+			for i := 0; i < attempts; i++ {
+				for j := range issuers.Status.Conditions {
+					if issuers.Status.Conditions[j].Status == v1.ConditionStatus(api.IssuerConditionReady) {
```

Comment:
Is this check working? In my test, this condition was never true, but it seemed to work fine even after the for loop was done with all 10 iterations. Seems like it should be much quicker than than in the normal case.

```
2022-04-05 22:47:13.665602 I | operator: setting up the controller-runtime manager
2022-04-05 22:47:13.970829 I | operator: Fetching webhook cert-manager-webhook to see if cert-manager is installed.
2022-04-05 22:47:13.974537 I | operator: Fetching Issuer rook-ceph/selfsigned-issuer.
2022-04-05 22:47:13.977380 I | operator: Creating Issuer rook-ceph/selfsigned-issuer.
2022-04-05 22:47:15.041117 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:16.042056 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:17.044357 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:18.045310 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:19.046770 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:20.047176 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:21.049249 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:22.050229 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:23.051190 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:24.052806 I | operator: waiting for webhook config "selfsigned-issuer" to be in ready status
2022-04-05 22:47:24.053401 I | operator: Fetching certificate rook-ceph/rook-admission-controller-cert.
2022-04-05 22:47:24.062834 I | operator: Creating certificate rook-ceph/rook-admission-controller-cert.
2022-04-05 22:47:24.089582 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:25.091257 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:26.092928 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:27.093586 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:28.098609 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:29.100724 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:30.102758 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:31.104970 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:32.105443 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:33.107559 I | operator: waiting for webhook config "rook-admission-controller-cert" to be in ready status
2022-04-05 22:47:34.107883 I | operator: looking for admission webhook secret "rook-ceph-admission-controller"
2022-04-05 22:47:34.115297 I | operator: admission webhook secret "rook-ceph-admission-controller" found
2022-04-05 22:47:34.115527 I | operator: Fetching webhook rook-ceph/rook-ceph-webhook.
2022-04-05 22:47:34.120102 I | operator: Creating webhook rook-ceph/rook-ceph-webhook.
2022-04-05 22:47:34.147804 I | operator: setting up admission webhooks
2022-04-05 22:47:34.150582 I | ceph-cluster-controller: successfully started
```

Reply:
After I restarted the operator everything was fast since the cert and issuer already existed.

Reply:
so we are good here right?

Reply:
My question is if the check inside the loop is working correctly. If it times out after all 10 iterations, I would expect the admission controller to fail starting. But since it succeeded to start, it seems the check should have succeeded sooner. Perhaps the check for the condition is not quite right?

Reply:
right, I have removed one of the loops which I think is not required